### ISSUE TITLE
[type] infinite stack when resolving recursive type parameters

### ISSUE THREAD

**Issue by timvandam** [deepkit/deepkit-framework#477]
```ts
class Abc<T> {}
export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };
function repro<T extends JsonValue>(type?: ReceiveType<T>) {
    type = resolveReceiveType(type);
    return function aaa(): Abc<T> {
        return new Abc();
    };
}

const fun = repro<{ type: 'aaa'; ya: 'bbb' }>();
```

Running `resolveReceiveType` eventually causes a V8 OOM

**Comment by marcj** [comment]
That's fixed

**Comment by marcj** [comment]
@alpharder good catch, thanks. indeed, some circular references weren't detected correctly. I've refactored the code to move the circular reference solver into the typeguard/serializer API itself so the various type templates don't have to take care of that. This should now support much more circular references, if not all. I've added your JSONValue type to our test suite to make sure it keeps working.

fixed in https://github.com/deepkit/deepkit-framework/commit/5f6bd124aaf9c546014b81dbded8110312f4e819
